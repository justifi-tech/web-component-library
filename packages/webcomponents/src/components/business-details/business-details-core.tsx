import { Component, Event, EventEmitter, Prop, State, h } from '@stencil/core';
import { Business } from '../../api/Business';
import { ErrorState } from '../../ui-components/details/utils';
import { ComponentError } from '../../api/ComponentError';
import { Skeleton, StyledHost } from '../../ui-components';

enum RENDER_STATES {
  LOADING = 'loading',
  READY = 'ready',
  ERROR = 'error',
}

@Component({
  tag: 'business-details-core'
})
export class BusinessDetailsCore {
  @Prop() getBusiness: Function;

  @State() business: Business;
  @State() renderState: RENDER_STATES = RENDER_STATES.LOADING;
  @State() errorMessage: string = 'An error ocurred.';

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  async componentWillLoad() {
    if (this.getBusiness) {
      this.fetchData();
    }
  }

  fetchData() {
    this.renderState = RENDER_STATES.LOADING;

    this.getBusiness({
      onSuccess: ({ business }) => {
        this.business = business;
        this.renderState = RENDER_STATES.READY;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.renderState = RENDER_STATES.ERROR;

        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
      }
    });
  }

  render() {
    if (this.renderState === RENDER_STATES.LOADING) {
      return (
        <StyledHost>
          <Skeleton height="100px" />
          <Skeleton height="100px" />
          <Skeleton height="100px" />
        </StyledHost>
      );
    }

    if (this.renderState === RENDER_STATES.ERROR) {
      return <StyledHost>{ErrorState(this.errorMessage)}</StyledHost>;
    }

    return (
      <StyledHost>
        <generic-info-details business={this.business} />
        <legal-address-details legalAddress={this.business.legal_address} />
        <representative-details representative={this.business.representative} />
        <owner-details owners={this.business.owners} />
        <additional-questions-details
          additionalQuestions={this.business.additional_questions}
        />
      </StyledHost>
    );
  }
}
