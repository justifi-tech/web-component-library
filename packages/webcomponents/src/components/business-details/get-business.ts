import { Business } from '../../api/Business';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetBusiness =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchBusiness(id, authToken);

      if (!response.error) {
        onSuccess({ business: new Business(response.data) });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError({ error: responseError });
      }
    } catch (error) {
      return onError({ error: error.message || error });
    }
  };
