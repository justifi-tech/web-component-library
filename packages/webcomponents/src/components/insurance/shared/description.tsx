import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-description',
  shadow: true,
})

export class InsuranceDescription {
  render() {
    <div></div>
  };
};