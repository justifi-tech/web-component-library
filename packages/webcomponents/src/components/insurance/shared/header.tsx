import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-header',
  shadow: true,
})

export class InsuranceHeader {
  render() {
    <div></div>
  };
};