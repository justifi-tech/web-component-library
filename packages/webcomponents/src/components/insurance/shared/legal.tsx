import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-legal',
  shadow: true,
})

export class InsuranceLegal {
  render() {
    <div>
      <slot></slot>
    </div>
  };
};