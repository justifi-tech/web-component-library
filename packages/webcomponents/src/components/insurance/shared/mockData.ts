export const quote = {
  "id": "4b154cc4-6f1a-47b3-bdd9-3d22c9027b24",
  "master_policy_id": "4b154cc4-6f1a-47b3-bdd9-3d22c9027b24",
  "partner": {
    "id": "a21c59ee-2fd9-45c0-b6b3-8402bc7c67ba",
    "legal_business_name": "Vertical Insure, Inc."
  },
  "policy_status": "QUOTED",
  "expiration_date": "2024-02-29",
  "effective_date": "2024-02-28",
  "quote_date": "2024-02-28T19:41:02.840256Z",
  "product": {
    "id": "716f9506-aeb1-4738-9b1a-6e48e90dc0f6",
    "name": "registration-cancellation",
    "friendly_name": "Registration Cancellation Insurance",
    "abbreviation": "RC",
    "description": "Spinnaker Registration Cancellation - No Work Coverage",
    "perils": [
      "Illness (including COVID-19)",
      "Injury",
      "Weather / Natural Disasters",
      "Employment",
      "Transportation",
      "Many other covered reasons"
    ],
    "promotional_header": "Add Registration Cancellation Protection",
    "promotional_description": "<p>Protect your registration and get your money back in the event of unforeseen circumstances that prevent participation, including sickness (including COVID-19), injury, inclement weather, transportation, job loss and other covered reasons.</p>",
    "legal_disclaimer": "<p>battleface Travel Insurance plans are underwritten by Spinnaker Insurance Company (an IL Corporation, NAIC# 24376), with administrative office at One Pluckemin Way, Suite 102, Bedminster, NJ 07921. Plans are offered and administered by battleface Insurance Services LLC, 45 East Lincoln Street, Columbus, OH 43215, National Producer Number 18731960 (FL License number L107363/CA License number 0M75381). <a rel=\"noopener noreferrer\" href=\"https://verticalinsure.com/s/Vertical-Insure-Disclosures-SPI-040123.pdf\" target=\"_blank\">Terms and conditions apply</a>. <a rel=\"noopener noreferrer\" href=\"https://www.battleface.com/en-us/spi-vertical-insure-library/\" target=\"_blank\">You can view the policy here</a>.</p>"
  },
  "policy_attributes": {
    "@type": "registration-cancellation",
    "carrier_product_code": "1010",
    "coverage_date": "DAY_OF",
    "participant": {
      "first_name": "John",
      "last_name": "Doe"
    },
    "insurable_amount": 50000,
    "event_start_date": "2024-02-29",
    "event_end_date": "2024-02-29"
  },
  "premium_amount": 3375,
  "currency": "USD",
  "total": 3375,
  "quote_id": "r1azwq4l0u9dzwnpd7dtu48t",
  "is_test": true,
  "checkout_url": "https://landing.verticalinsure.com/verticalinsure/registration-cancellation/r1azwq4l0u9dzwnpd7dtu48t?id=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ0ZXN0MDFAdmVydGljYWxpbnN1cmUuY29tIiwiZW52aXJvbm1lbnQiOiJURVNUIiwicmVzb3VyY2UiOiJyMWF6d3E0bDB1OWR6d25wZDdkdHU0OHQiLCJwZXJtaXNzaW9ucyI6WyJQVVJDSEFTRV9QT0xJQ1kiXSwiaXNzIjoiaHR0cHM6Ly92ZXJ0aWNhbGluc3VyZS5jb20iLCJleHAiOjE3MTAzNTk3MzUsImlhdCI6MTcwOTE1MDEzNX0.ICLtb6MLQ9LuNOZi_2XYiPpt3ggiUoN-nYbEchL7ZpA",
  "policy_holder": {
    "email_address": "test01@verticalinsure.com",
    "country": "US",
    "id": "20c1d250-89d0-3904-84fc-5ee46cf2cee3",
    "state": "MN"
  }
};
