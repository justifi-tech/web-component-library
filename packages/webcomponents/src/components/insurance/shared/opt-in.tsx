import { Component, h } from '@stencil/core';

@Component({
  tag: 'justifi-insurance-opt-in',
  shadow: true,
})

export class InsuranceOptIn {
  render() {
    <div></div>
  };
};