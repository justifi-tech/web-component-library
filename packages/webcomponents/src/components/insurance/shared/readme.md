# justifi-insurance-opt-in



<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [justifi-insurance-registration-cancellation](../registration-cancellation)

### Graph
```mermaid
graph TD;
  justifi-insurance-registration-cancellation --> justifi-insurance-opt-in
  style justifi-insurance-opt-in fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
