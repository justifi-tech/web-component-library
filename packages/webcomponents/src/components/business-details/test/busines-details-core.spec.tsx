jest.mock('../../../utils/styled-host/modified-bootstrap.css', () => '');

import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";
import { BusinessDetailsCore } from "../business-details-core";
import { makeGetBusiness } from "../get-business";
import mockBusinessDetails from './../../../../../../mockData/mockBusinessDetails.json';
import { Business, IBusiness } from "../../../api/Business";
import { API_NOT_AUTHENTICATED_ERROR } from "../../../api/shared";

describe('BusinessDetailsCore', () => {
  it('should display loading state correclty', async () => {
    const page = await newSpecPage({
      components: [BusinessDetailsCore],
      template: () => <business-details-core />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should display error state correctly', async () => {
    const getBusiness = makeGetBusiness({
      id: '',
      authToken: '',
      service: {
        fetchBusiness: async () => ({ error: 'error' }),
      },
    });

    const page = await newSpecPage({
      components: [BusinessDetailsCore],
      template: () => <business-details-core getBusiness={getBusiness} />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should set business details correctly to state', async () => {
    const getBusiness = makeGetBusiness({
      id: '',
      authToken: '',
      service: {
        fetchBusiness: async () => Promise.resolve(mockBusinessDetails),
      },
    });

    const page = await newSpecPage({
      components: [BusinessDetailsCore],
      template: () => <business-details-core getBusiness={getBusiness} />,
    });

    await page.waitForChanges();

    const expectedBusiness = new Business(mockBusinessDetails.data as unknown as IBusiness);
    expect(page.rootInstance.business).toEqual(expectedBusiness);
    expect(page.root).toMatchSnapshot();
  });

  it('should emit error event when fetch fails', async () => {
    const getBusiness = makeGetBusiness({
      authToken: 'abc',
      id: 'abc',
      service: {
        fetchBusiness: jest.fn().mockRejectedValue(new Error('Fetch error')),
      },
    })

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [BusinessDetailsCore],
      template: () => <business-details-core getBusiness={getBusiness} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Fetch error',
          severity: 'error',
        }
      })
    );
  });

  it('should emit error event when API returns error', async () => {
    const getBusiness = makeGetBusiness({
      authToken: 'abc',
      id: 'abc',
      service: {
        fetchBusiness: jest.fn().mockResolvedValue(API_NOT_AUTHENTICATED_ERROR),
      },
    })

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [BusinessDetailsCore],
      template: () => <business-details-core getBusiness={getBusiness} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'not-authenticated',
          message: 'Not Authenticated',
          severity: 'error',
        }
      })
    );
  });
});

