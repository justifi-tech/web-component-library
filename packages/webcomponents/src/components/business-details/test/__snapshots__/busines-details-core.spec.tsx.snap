// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessDetailsCore should display error state correctly 1`] = `
<business-details-core class="body">
  <style></style>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
    <div>
      error
    </div>
  </main>
</business-details-core>
`;

exports[`BusinessDetailsCore should display loading state correclty 1`] = `
<business-details-core class="body">
  <style></style>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-loading-state" style="font-size: 1.2rem;">
    <div class="spinner-border spinner-border-sm" part="detail-loading-spinner" role="status">
      <span class="visually-hidden">
        Loading...
      </span>
    </div>
  </main>
</business-details-core>
`;

exports[`BusinessDetailsCore should set business details correctly to state 1`] = `
<business-details-core class="body">
  <style></style>
  <generic-info-details></generic-info-details>
  <legal-address-details></legal-address-details>
  <representative-details></representative-details>
  <owner-details></owner-details>
  <additional-questions-details></additional-questions-details>
</business-details-core>
`;
